dot/state/grandpa.go

```diff
@@ -26,42 +37,477 @@ var (
+	}
+
+	pendingChange := &pendingChange{
+		bestFinalizedNumber: fc.BestFinalizedBlock,
+		nextAuthorities:     auths,
+		announcingHeader:    header,
+		delay:               fc.Delay,
+	}
+
+	s.forcedChanges = append(s.forcedChanges, pendingChange)
+	sort.Sort(s.forcedChanges)
+
+	logger.Debugf("there are now %d possible forced changes", len(s.forcedChanges))
+	return nil
+}
+
+func (s *GrandpaState) addScheduledChange(header *types.Header, sc types.GrandpaScheduledChange) error {
+	auths, err := types.GrandpaAuthoritiesRawToAuthorities(sc.Auths)
+	if err != nil {
+		return fmt.Errorf("cannot parse GRANPDA authorities to raw authorities: %w", err)
+	}
+
+	pendingChange := &pendingChange{
+		nextAuthorities:  auths,
+		announcingHeader: header,
+		delay:            sc.Delay,
+	}
+
+	return s.importScheduledChange(pendingChange)
+}
+
+func (s *GrandpaState) importScheduledChange(pendingChange *pendingChange) error {
+	for _, root := range s.scheduledChangeRoots {
+		imported, err := root.importScheduledChange(pendingChange.announcingHeader.Hash(),
+			pendingChange.announcingHeader.Number, pendingChange, s.blockState.IsDescendantOf)
+
+		if err != nil {
+			return fmt.Errorf("could not import scheduled change: %w", err)
+		}
+
+		if imported {
+			logger.Debugf("changes on header %s (%d) imported successfully",
+				pendingChange.announcingHeader.Hash(), pendingChange.announcingHeader.Number)
+			return nil
+		}
+	}
+
+	pendingChangeNode := &pendingChangeNode{
+		change: pendingChange,
+		nodes:  []*pendingChangeNode{},
+	}
+
+	s.scheduledChangeRoots = append(s.scheduledChangeRoots, pendingChangeNode)
+	logger.Debugf("there are now %d possible scheduled changes (roots)", len(s.scheduledChangeRoots))
+
+	return nil
+}
+
+// getApplicableScheduledChange iterates through the scheduled change tree roots looking
+// for the change node which contains a lower or equal effective number, to apply.
+// When we found that node we update the scheduled change tree roots with its children
+// that belongs to the same finalized node branch. If we don't find such node we update the scheduled
+// change tree roots with the change nodes that belongs to the same finalized node branch
+func (s *GrandpaState) getApplicableScheduledChange(finalizedHash common.Hash, finalizedNumber uint) (
+	change *pendingChange, err error) {
+
+	var changeNode *pendingChangeNode
+	for _, root := range s.scheduledChangeRoots {
+		if root.change.effectiveNumber() > finalizedNumber {
+			continue
+		}
+
+		changeNodeHash := root.change.announcingHeader.Hash()
+
+		// if the change doesn't have the same hash
+		// neither the finalized header is descendant of the change then skip to the next root
+		if !finalizedHash.Equal(changeNodeHash) {
+			isDescendant, err := s.blockState.IsDescendantOf(changeNodeHash, finalizedHash)
+			if err != nil {
+				return nil, fmt.Errorf("cannot verify ancestry: %w", err)
+			}
+
+			if !isDescendant {
+				continue
+			}
```

Comment:
The comment is unclear, especially with the `neither`, can you clarify this please 🤔